### PR TITLE
Creating a V2, which clearly places emphasis on risk over intention

### PR DESCRIPTION
**Pull request by arlobelshee**
This is a backwards-incompatible change.

I've been observing for a while that most users of the current notation think of the intention first, and then hand-wave the risk level. My goal for even using this notation is to become aware of risk on a commit-by-commit level. Thus I want to separate the moment where someone thinks / labels risk from the moment when they label intention.

Since I'm breaking compatibility anyway, now seemed a good time to ensure we don't have character sequences that cause conflicts. I'm trying to avoid clashes with:

* Markdown
* Command-lines (bash, pwsh, or command)
* Comments in common programming languages

I *think* this proposal avoids all of those, but please help me make it so. Also, are there other clashes we should worry about? Where else will commit messages appear?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,16 +15,26 @@ This is particularly useful when:
...
+| **Known safe** | lowercase letter | `r - Extract method Applesauce` | Addresses all known and unknown risks. | Intended Change, Known Invariants, Unknown Invariants |
+| **Validated** | uppercase letter | `R - Extract method Applesauce` | Addresses all known risks. | Intended Change, Known Invariants |
+| **Risky** | uppercase followed by 2 bangs | `R!! Extract method Applesauce` | Some known risks remain unverified. | Intended Change |
+| **(Probably) Broken** | uppercase followed by 2 stars | `R** Start extracting method with no name` | No risk attestation. |  |
+
+Behavior categories:
+
+* **Intended Change:** The 0 or 1 behavior change intended in the commit. Could be verified by one test assertion. By default, a commit with more than 1 behavior change cannot be represented at any risk level below *Probably Broken*.
+* **Known Invariants:** All behaviors known to the development team at the time the change was made. Automated tests can greatly increse the size of this set, thus enhancing safety when commits are at a risk level that guarantees correctness for Known Invariants. However, this set also includes behaviors that are known but not tested.
```
Comment (JayBazuzi) [inline README.md:30]: spelling: increase

**Comment by JayBazuzi** [comment]
I thought maybe pound `#` rather than at `@` might stand out a little better, and fit the vertical / horizontal lines motif. Did you consider that?

**Comment by JayBazuzi** [comment]
I will bring this new design to my teams to try out.

**Comment by arlobelshee** [comment]
I considered #, but was worried about it becoming a comment in a python
context. I do think it stands out better and fits the theme well.

I also considered +, but that that is a markdown list. @ seems to only be a
special char when followed by non whitespace.

I'm not sure which of these conflicts I should actually worry about.

As a python programmer, are you worried about the comment char overlap?

Arlo

**Comment by JayBazuzi** [comment]
I am not worried about programming languages' comment characters because this would always be in a quoted string. Is there an in-code use case I'm missing? 

I am imagine interactive command line as being the primary use case, so I think about shell special characters first: backslash, backtick, ampersand, precent, dollar, etc. I think we're good there.

Markdown for writing docs about using the notation is a minor problem for `#`, like this:

# R clean up the code and make it faster

But that is easily/appropriately addressed with a code block:

```
# R clean up the code and make it faster
```

**Comment by arlobelshee** [comment]
My concern with markdown is that many Web tools display all commit comments
in a markdown context. They explicitly support the use of markdown in the
comment. So it often would not appear inside a code fence.

**Comment by JayBazuzi** [comment]
Ah, yeah: if a merge commit is automatically populated with the list of commits in the source branch, that might be all formatted at markdown. Makes sense to avoid leading `#` for that reason.

**Comment by JayBazuzi** [comment]
I made a migration guide. https://github.com/RefactoringCombos/ArlosCommitNotation/blob/738f0b074b4ff2e8786ae9071688024c2baf141e/Migrating%20from%20V1%20to%20V2.md

File: README.md [hunk 2]
```diff
@@ -86,13 +102,15 @@ A Refactoring or Remodeling intends to alter the program in some way without cha
...
 | Code | Known Approaches |
 | --- | --- |
-| `r - ` | One of: <ul><li>Provable refactoring<sup>[2]</sup></li><li>Test-supported Procedural Refactoring<sup>[3]</sup> entirely within test code</li></ul> |
-| `R - ` | Test-supported Procedural Refactoring<sup>[3]</sup> |
-| `R!!` | Identified single, named refactoring, but executed by editing code or without whole-project test coverage. |
-| `R**` | Remodeled by editing code, even in small chunks. |
+| `. R` | One of: <ul><li>Provable refactoring<sup>[2]</sup></li><li>Test-supported Procedural Refactoring<sup>[3]</sup> entirely within test code</li></ul> |
+| `- R` | Test-supported Procedural Refactoring<sup>[3]</sup> |
+| `! R` | Identified single, named refactoring, but executed by editing code or without whole-project test coverage. |
+| `* R` | Remodeled by editing code, even in small chunks. |
```
Comment (7pluts23) [inline README.md:108]: The switch from * to @ was missed on this line

**Comment by JayBazuzi** [comment]
A team I work with tried out ACN V2. Their feedback:

- they like intention all being uppercase, as it's less to think about.

- they like risk coming first, as it forces them to ask "how safe is this?" without getting distracted by intention. Previously they fell in to the trap of "this is a CI change, that's e" without noticing that it could break production.

File: README.md (hunk 2 above)
Comment (arlobelshee) [reply README.md]: Fixed. Thanks!

**Comment by arlobelshee** [comment]
The proposal now uses the following risk codes:

| Code | Meaning |
| --- | --- |
| `.` | Bug-for-bug compatible |
| `^` | No potentially-preventable regressions; may have regressed behaviors our customers use but we aren't aware of |
| `!` | The intended change was correct; may have impacted other functionality |
| `@` | No promises! It might work, or not. |

**Comment by arlobelshee** [comment]
I considered other risk codes. As per discussion with Llewellyn and Jay, we wanted to get `!` to be the second level, so that people learning to refactor feel a strong desire to keep it provably safe. I'm not sure whether `^` will elicit the same pain as `!` does.

That said, the second level is the "best" level for features and bug fixes. So I want it to feel "nice" in that case. Also, most of the characters that are more explosive than `!` have special meaning in either the shell or markdown.

Here are other things I considered. Do any of these stand out to you as good, or do you have an alternate suggestion?

| Current | Proposed |
| --- | --- |
| `.` | `.` |
| `^` | `!` |
| `!` | `$` `%` `&` or `/` |
| `@` | `@` |

`$`, `%`, and `&` are only OK in the shell because the will be followed by a space and included in some kind of quotes. Still, this seems very error-prone, so I am leaning against it.

The primary other alternative would be `#`, but that has meaning in markdown when used at the start of a line followed by space. That's the same reason we aren't using `-`, `+`, or `*` as risk codes.

**Comment by JayBazuzi** [comment]
I particularly dislike `&` because it's not symmetrical. `/` is asymmetric but less so; `@` is close enough to symmetrical for me.

`$` isn't as accessible on a lot of non-us keyboards, but programmers worldwide have figured out how to type this character thanks to Bash variables.

**Comment by JayBazuzi** [comment]
Did you consider `|`? Not sure where I'd put that.

I am suddenly reminded of ASCII wait animation, cycling `|/-\`.

**Comment by arlobelshee** [comment]
I considered pipe, but that also has a high chance of command line problem.

**Comment by JayBazuzi** [comment]
`.` / `o` / `O` / `@` is an interesting visual sequence.

**Comment by arlobelshee** [comment]
Any votes between the following 2 options?

| Option A | Option B | Meaning |
| --- | --- | --- |
| `.` | `.` | Bug-for-bug compatible |
| `^` | `o` | No potentially-preventable regressions; may have regressed behaviors our customers use but we aren't aware of |
| `!` | `O` | The intended change was correct; may have impacted other functionality |
| `@` | `@` | No promises! It might work, or not. |

**Comment by JayBazuzi** [comment]
> Any votes between the following 2 options?
> ...

I asked around and got this reply:

> I sort of like Option B. In terms of code review, `@` notation would stand out far better in a sea of 0 or o than the other notations in Option A. Also I feel like Option B takes less of my memory capacity.

For me personally, having middle pair be similar to each other is quite a different cast than the low pair / high pair.

**Comment by arlobelshee** [comment]
Listening to your 2 responses above, Jay, and playing with it myself, I'm liking the first 2 being little circles. And I want the last 2 to stand out. So I think that really settles all levels except the third. I think we can choose between these two options as our final choice.

Any votes between the following 2 options?

| Option A | Option B | Meaning |
| --- | --- | --- |
| `.` | `.` | Bug-for-bug compatible |
| `o` | `o` | No potentially-preventable regressions; may have regressed behaviors our customers use but we aren't aware of |
| `!` | `%` | The intended change was correct; may have impacted other functionality |
| `@` | `@` | No promises! It might work, or not. |

**Comment by arlobelshee** [comment]
I wrote a response and then deleted it. I'm trying to figure out a nuance of my intent.

I think it comes down to a question: when reviewing a sea of changes, do you want well-done feature changes to fade into the background or stand out?

This determines whether we want the visual filter to be between levels 1/2 or between 2/3. In other words, do we want to "see" levels 2, 3, 4 (ignore only level 1), or do we want to "see" levels 3, 4 (ignore levels 1 & 2)?

Interestingly, using the `o` for level 2 gives us both options, depending on what character we use for level 3. Here are 3 options. I feel that Option A brings the top 3 levels into the visual foreground, while Options B & C only bring the top 2 levels into the visual foreground.

So should level 2 be in the foreground or background? And which option best gives that result?

| Option A | Option B | Option C | Meaning |
| --- | --- | --- | --- |
| `.` |  `.` | `.` | Bug-for-bug compatible |
| `o` |  `o` | `o` | No potentially-preventable regressions; may have regressed behaviors our customers use but we aren't aware of |
| `O` |  `!` | `%` | The intended change was correct; may have impacted other functionality |
| `@` | `@` | `@` | No promises! It might work, or not. |

**Comment by JayBazuzi** [comment]
I want both scanning and discomfort, and which levels I care about varies by context.

As a code reviewer looking at a PR I want to quickly identify level 3 and 4 commits so I can scrutinize them for mistakes or give feedback to the developer about reworking the change (both 4->3 rework and 3->2/1 rework are improvements I want to encourage).

As a spelunker trying to figure out why the last release broke production I want to find all level 2, 3, and 4 commits because that's where the risk is. 

As a release engineer I want to find the first 2/3/4 commit and maybe ship everything before that point in one go with minimal ceremony, and then I can follow up with more ceremony to ship the rest. 

When committing a change I want us to feel discomfort when we're using a higher risk level than we need to.
- We can go from 4 to 3 by breaking up changes and including tests. We should almost always do that.
- Building out the test coverage (and modularity) to go from 3 to 2 is a big job that doesn't easily slip into the current User Story but if the test coverage is in place then we should be able to hit 2 most of the time.
- 1 is generally only possible if we're working in a language + tooling that supports it. Switching to a proper language will not fit in the current User Story. If we're at 2 we have probably done as well as we can today.

So there are cases where 1/2 is the right boundary, where 2/3 is the right boundary, and 3/4 is the right boundary.

While writing release notes I want to see all the `F` and `B` changes.

I see we're using numbers in this discussion. Last week a dev on one of my teams said he wants the risk level to be digits 1..9, instead of inventing new symbols. Obviously, that loses scanability but I wanted to pass it on.

This reminds me that I wish we had proper tooling. I wish that Git had a general-purpose commit metadata system beyond author, time, tag, signature - I want to add a Risk tag and an Intent tag and then I want a log view that I can filter and highlight. Maybe one day.

**Comment by JayBazuzi** [comment]
Discussed with Emily Bache this morning and she said that `o` / `O` look too much alike to discern, even when you're looking closely.

File: README.md [hunk 3]
```diff
@@ -50,15 +62,17 @@ Each intention can appear at any of the 4 risk levels. Each intention's full det
 
 | Code | Known Approaches |
 | --- | --- |
-| `f - ` | None known |
-| `F - ` | Meets all of:<ul><li>Change is <= 8 <abbr title="lines of code">LoC</abbr><sup>[5]</sup></li><li>Feature was fully unit tested prior to this change.</li><li>Change includes new or changed unit tests to match intended behavior alteration.</li></ul> |
-| `F!!` | Change includes unit tests for new behavior. |
-| `F**` | No automatic tests, or unfinished implementation. |
+| `. F` | Meets all criteria for `- F` and developers are the only users of the feature. For example, extends build tooling for your own build or adds debug logging. |
```
Comment (raymyers) [inline README.md:65]: In this version, I'm not seeing a `- F`. Does this refer to the requirements of `^ F`?

File: README.md [hunk 4]
```diff
@@ -15,16 +15,26 @@ This is particularly useful when:
...
-
-* **Known safe:** Developer performed the task in a way that prevents the potential risks, even for situations that developer is not aware of.
-* **Validated:** Developer performed the task in some way that includes validation for all risks the developer thought of. The most common technique is developer-written automated tests.
-* **Risky:** Developer is aware of risks and attempted to mitigate them as much as possible, but there is no formal verification. Commonly this includes a manual change that the developer could not fully verify.
+We divide all behaviors of the system into 3 sets. The change is intended to alter the *Intended Change* while not altering any of the *Invariants*. The *Risk Levels* are based on correctness guarantees: which invariants can this commit guarantee did not change, and can this commit guarantee that it changed the intended change in the way the authors intended?
+
+| Risk Level | Code | Example | Meaning | Correctness Guarantees |
+| --- | --- | --- | --- | --- |
+| **Known safe** | `.` | `. r Extract method` | Addresses all known and unknown risks. | Intended Change, Known Invariants, Unknown Invariants |
```
Comment (MichaelRWolf) [inline README.md:22]: Suggest: Parallelism -- "(Known) Safe", preserving the strong intent of provably safe (like "(Probably) Broken"), and also allowing reference to unambiguous, single words for each level -- Safe, Validated, Risky, Broken.

File: README.md [hunk 5]
```diff
@@ -37,10 +47,12 @@ Each intention can appear at any of the 4 risk levels. Each intention's full det
 | --- | --- | --- |
 | F | Feature | Change or extend one aspect of program behavior without altering others. |
 | B | Bugfix | Repair one existing, undesirable program behavior without altering any others. |
-| R | Refactoring | Change implementation without changing program behavior. |
-| D | Documentation | Change something which communicates to team members and does not impact program behavior. |
+| r | Refactoring | Change implementation without changing program behavior. |
+| d | Documentation | Change something which communicates to team members and does not impact program behavior. |
```
Comment (MichaelRWolf) [inline README.md:51]: Capitalization is inconsistent.  Is this intentional?
- Yes -- Explain why.
- No -- Make it consistent.

**Review by MichaelRWolf — COMMENTED** [review]
There is a trend in the medical community to move away from naming diseases after individuals and instead adopt more descriptive and systematic names.  Doing so here would relax the sense of individual ownership by Arlo to a more community-owned sense for the practice.  

Suggestions:

- RICN - Risk/Intent Change Notation (strikes balance between old- and new-school tools)
- RICN - Risk/Intent Commit Notation (too new-school-specific)
- RICN - Risk/Intent Checkin Notation (too old-school-specific)

In addition, _front-loading_ the name with "Risk" further emphasizes the intent for both the _notation_ and the _thinking_ and _practice_ behind the notation.

**Comment by arlobelshee** [comment]
So Jay, what is your preferred indication for each risk level?

I've been doing ./o/O/@, but am very happy to adopt another alternative.

Arlo

**Comment by MichaelRWolf** [comment]
Have you looked at any of the (decades-old by now) ASCII art generators, or translators?  They mostly map to lines and angles, but some of the more interesting ones had grey-scale mappings that may help with the back-of-the-retina precessing.  But I also bet that they are not portable.

BTW... I have been playing with a message-generating script (tentatively called RIMC - Risk/Intent Commit Message) that hard-codes ./^/!/@, but I had intended to have that mapping pulled from a config file, so everyone could create their own mapping (and thus defeat one of the intents of having a "standard").

Once I've got the 'encoder' working, I had toyed around with creating a 'decoder' that would translate "standard" messages by adding color.  The ANSI escape sequences are at least "standard" and implemented on terminals (and emulators) for about 50 years.

**Comment by MichaelRWolf** [comment]
I'm going to propose using it in some public mobs.

I've creating messages like this...

```
$ (./ricm.py --risk risky --intent feature 'Add --HELP_FOR_RICM to display risk and intent options'; echo; cat contributers) | pbcopy; pbpaste
! f Add --HELP_FOR_RICM to display risk and intent options

Manny
Moe
Jack
```

Then either (1) paste into commit message of GUI (IntelliJ, GitKraken, etc...) or (2) recall this line from history in a shell
```
git commit -m "$(pbpaste)"
```

**Comment by JayBazuzi** [comment]
> So Jay, what is your preferred indication for each risk level?

I have been trying both and I don't have a definitive answer. Preference feelings that come up for me are mild and they don't clump up around one option.

If you said that the only way that V2 could ship is if Jay made a choice, I would pick ./o/O/@.

File: README.md (hunk 5 above)
Comment (JayBazuzi) [reply README.md:51]: Uppercase = intended to change behavior.

Lowercase = intended to preserve behavior.

**Comment by JayBazuzi** [comment]
I've continued to use ./o/O/@ and today got some feedback that `o r` looks too much like the English word "or".

However, this came from someone who generally doesn't use ACN except for F**/B**/R**, so I don't know if they're in a good position to judge the system.

I'm now experimenting with surround with brackets, e.g. `[o r]` to see how that feels.

**Comment by JayBazuzi** [comment]
We could offer two options. "Experiment with these with your team and choose the one that suits people better."

| Tag     | Risk                               |
|---------|------------------------------------|
| `.`     | Known Safe                         |
| `o`/`^` | Addresses all known risks          |
| `O`/`!` | Some known risks remain unverified |
| `@`     | No risk attestation                |

File: README.md (hunk 5 above)
Comment (MichaelRWolf) [reply README.md:51]: Ah... understood.

Could this be added as a key to the table?


UPPERCASE = intended to change behavior
lowercase = intended to preserve behavior

**Comment by JayBazuzi** [comment]
`-` looks like a very natural punctuation in certain contexts, which detracts from its effectiveness as a commit notation. For example here it looks like it just separates the commit ID from its description.

image

File: Learning Path.md [hunk 6]
```diff
@@ -43,102 +43,102 @@ Refactorings often have a large diff even though they don't change behavior and
...
 ## 3. Pick up easy wins
 
 Once the team has learned the above practices, the following tags are easy to adopt.
 
-- `t` for test-only changes
-- `a` for auto-formatting
-- `R!!` for named refactorings
+- `. t` for test-only changes
+- `. a` for auto-formatting
+- `! R` for named refactorings
```
Comment (JayBazuzi) [inline Learning Path.md:58]:
```suggestion
- `! r` for named refactorings
```

File: Learning Path.md [hunk 7]
```diff
@@ -43,102 +43,102 @@ Refactorings often have a large diff even though they don't change behavior and
...
 
 ### Example commit history
 
 ```
-t - fill in missing tests for existing login code
-R!! Extract Method
-a - autoformat with prettier
+. t - fill in missing tests for existing login code
+! R Extract Method
```
Comment (JayBazuzi) [inline Learning Path.md:64]:
```suggestion
! r Extract Method
```

File: README.md (hunk 5 above)
Comment (arlobelshee) [reply README.md:51]: That was the intention, and now I'm thinking to go a different way: allow F or f for feature, with F meaning this is user-visible, and f meaning it changes user-invisible functionality. Can be used to decide what should appear in release notes / changelog. I'm going to provide for 2 variations here, by letting the team choose the meaning of capitalization.

**Comment by JayBazuzi** [comment]
Yay!

**Comment by Chris-May** [comment]
It’s been so great to watch this go through so much thought. Congrats!

**Pull request merged**